File: src/dart/main/Occur.dart
part of lingukit;

class Occur {
  
  /**
   * { min, max }
   */
  Occur.occur( int min, int max ) : this (min, max);
  
  Occur.x(int times) : this(times, times);
  
  static final Occur never = new Occur(0, 0);
  static final Occur once = new Occur(1 , 1);
  
  /**
   * *
   */
  static final Occur star = new Occur(0 , 1000);
  
  /**
   * +
   */
  static final Occur plus = new Occur(1 , 1000);
  
  /**
   * ?
   */
  static final Occur qmark = new Occur(0 , 1);

  Occur(this.min, this.max);

  final int min;
  final int max;
  
  String toString() {
    if (min == star.min && max == star.max) {
      return "*";
    }
    if (min == plus.min && max == plus.max) {
      return "+";
    }
    if (min == qmark.min && max == qmark.max) {
      return "?";
    }
    if (min == max) {
      return "x"+min.toString();
    }
    if (max == plus.max) {
      return "x"+min.toString()+"+";
    }
    return "x"+min.toString()+"-"+max.toString()+"";
  }
}

File: src/java/main/bruno/lang/grammar/Occur.java
package bruno.lang.grammar;

public final class Occur {
	
	public static final int MAX_OCCURANCE = 1000;

	/**
	 * <pre>
	 * { min, max }
	 * </pre>
	 */
	public static Occur occur( int min, int max ) {
		return new Occur(min, max);
	}
	
	public static Occur x(int times) {
		return occur(times, times);
	}
	
	public static final Occur never = occur(0, 0);
	
	
	/**
	 * <pre>
	 * *
	 * </pre>
	 */
	public static final Occur once = occur(1 , 1);
	
	/**
	 * <pre>
	 * *
	 * </pre>
	 */
	public static final Occur star = occur(0 , MAX_OCCURANCE);
	
	/**
	 * <pre>
	 * +
	 * </pre>
	 */
	public static final Occur plus = occur(1 , MAX_OCCURANCE);
	
	/**
	 * <pre>
	 * ?
	 * </pre>
	 */
	public static final Occur qmark = occur(0 , 1);

	public final int min;
	public final int max;

	Occur( int min, int max ) {
		super();
		this.min = min;
		this.max = max;
	}
	
	@Override
	public String toString() {
		if (min == star.min && max == star.max) {
			return "*";
		}
		if (min == plus.min && max == plus.max) {
			return "+";
		}
		if (min == qmark.min && max == qmark.max) {
			return "?";
		}
		if (min == max) {
			return "**"+min;
		}
		if (max == MAX_OCCURANCE) {
			return "**"+min+"+";
		}
		return "**"+min+".."+max+"";
	}

}